docs/architecture/adr-015-ibc-packet-receiver.md

```diff
@@ -0,0 +1,67 @@
+# ADR 015: IBC Packet Receiver
+
+## Changelog
+
+- 2019 Oct 22: Initial Draft
+
+## Context
+
+[ICS 26](https://github.com/cosmos/ics/tree/master/spec/ics-026-routing-module) defines function [`handlePacketRecv`](https://github.com/cosmos/ics/tree/master/spec/ics-026-routing-module#packet-relay). 
+`handlePacketRecv` executes per-module `onRecvPacket` callbacks, verifies the packet merkle proof, and pushes the acknowledgement bytes, if present,
+to the state. 
+
+The mechanism is similar to the transaction handling logic in `baseapp`. After authentication, the handler is executed, and 
+the authentication state change must be committed regardless of the result of the handler execution. 
+
+ICS 26 also requires acknowledgement writing which has to be done after the handler execution and also must be commited 
+regardless of the result of the handler execution.
+
+## Decision
+
+We will define an `AnteHandler` for IBC packet receiving. The `AnteHandler` will iterate over the messages included in the 
+transaction, type switch to check whether the message contains an incoming IBC packet, and if so verify the Merkle proof.
+
+```go
+// Pseudocode
+func IBCAnteHandler(ctx sdk.Context, tx sdk.Tx, bool) (sdk.Context, sdk.Result, bool) {
+  for _, msg := range tx.GetMsgs() {
+    if msg, ok := msg.(MsgPacket); ok {
+      if err := VerifyPacket(msg.Packet, msg.Proofs, msg.ProofHeight); err != nil {
+         return ctx, err.Result(), true
+      }
+    }
+  }
+  return ctx, sdk.Result{}, false
+}
+```
+
+where `MsgPacket` is the `sdk.Msg` type including any IBC packet inside and embedding `Packet.Route()` method.
+
+The `AnteHandler` will be inserted to the top level application, after the signature authentication logic provided by `auth.NewAnteHandler`, utilizing `AnteDecorator` pattern.
+
+This proposal does not cover automatic acknowledgement processing. Applications who want to return acknowledgement should manually handle multistore cacheing.
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+- Intuitive interface for developers - IBC handlers do not need to care about IBC authentication
+- State change commitment logic is embedded into `baseapp.runTx` logic
+- IBC applications do not need to define message type for receiving packet
+
+### Negative
+
+- AnteHandler processes over the whole transaction before the each handler, `UpdateClient` and `RecvPacket` cannot be processed in the same transaction
+- Cannot support dynamic ports, routing is tied to the baseapp router
```

Comment:
This is a serious problem. We will want dynamic port support, at least in the sense that ports can be bound at runtime - say for CosmWASM contracts which bind ports.

I also don't entirely understand why dynamic ports would be impossible with this design. We might need to write a different router, that's true.

Reply:
The messages embed the packets which have the method `Route()`. After `IBCAnteHandler` verifies the proof, the packet is then routed to the respective handlers using `baseapp.Router`, which is fixed at the initialization time. 

The problem with the dynamic port support on the sdk is that it is not secured by consensus because it is stored in memory. If we have a `rootmultistore` like structure for port allocation, it will still add and remove `pointer => portid` inside the memory. 

Afaik the intention was to keep every mutable state inside the tree instead of memory to make it falsifiable via apphash in the consensus, and having any mutability inside keeper related struct was considered antipattern. 

I am looking into this problem and trying to find a solution from swingset, but I am not sure if there is any safe way we can use golang pointers as capability key while keeping it consensus safe.

Reply:
One of the solutions comes into my mind is hierarchical ports: port is not defined by a single string but a list of strings(like keypath for nested multistore). Except for the last port id all the former ones can be static and not be modified in runtime(like rootmultistore). The last one should be able to modified in runtime(like iavlstore). The list of portids as a whole will be therefore runtime modifiable.

The last one will not use golang pointer as capability key but will define their own. In the example of ethermint module, the module has total control over the contracts, so it can track the capability of each contract without relying on golang pointer.

Reply:
I think we will need a 'dynamic capability' module or store in the SDK to provide dynamic capability keys, yes. This probably merits a separate ADR.

However, I still think the underlying implementation can be Golang pointers - we just need to recreate unique pointers for all capabilities whenever the blockchain is 'restarted' (e.g. `gaiad start` after stopping it, syncing from genesis, etc).

Reply:
Split into https://github.com/cosmos/cosmos-sdk/issues/5290.

Reply:
^^

Please write the summary of the comments and the reference to the new issue.

Reply:
^^